### PR TITLE
[WASI-NN] piper: add backend structure

### PR DESCRIPTION
**Pull request by PeterD1524**
This is the pull request for [LFX Workspace: Support piper as a new backend of the WASI-NN WasmEdge plugin #3483](https://www.github.com/WasmEdge/WasmEdge/issues/3483).

It will support Piper 2023.11.14-2 as a new WASI-NN backend:

- adapted the Piper command line program into WASI-NN API
- added a new backend index 11 for Piper according to [WASI-NN: Add new backends and discuss the corresponding index #3491](https://www.github.com/WasmEdge/WasmEdge/issues/3491)

### CODE REVIEW

**Comment by juntao** [comment]
Hello, I am a [code review bot](https://github.com/flows-network/github-pr-summary/) on [flows.network](https://flows.network/). Here are my reviews of code commits in this PR.

------

File: plugins/wasi_nn/piper.cpp [hunk 1]
```diff
@@ -0,0 +1,487 @@
...
+          GraphRef.Config->PhonemeSilenceSeconds;
+    } else {
+      // Merge
+      for (const auto &[Phoneme, SilenceSeconds] :
+           *GraphRef.Config->PhonemeSilenceSeconds) {
+        GraphRef.Voice->synthesisConfig.phonemeSilenceSeconds->try_emplace(
+            Phoneme, SilenceSeconds);
+      }
+    }
+
```
Comment (hydai) [inline plugins/wasi_nn/piper.cpp:285]: Redundant empty line.

File: plugins/wasi_nn/piper.cpp [hunk 2]
```diff
@@ -0,0 +1,487 @@
...
+template <typename T, typename U = T>
+Expect<WASINN::ErrNo> getOptionalOption(simdjson::dom::element &Doc,
+                                        std::string_view Key,
+                                        std::optional<T> &Result) {
+  if (Doc.at_key(Key).error() == simdjson::SUCCESS) {
+    auto Value = U{};
+    auto Err = Doc[Key].get<U>().get(Value);
+    if (Err) {
+      spdlog::error(
+          "[WASI-NN] Piper backend: Unable to retrieve the {} option.", Key);
```
Comment (hydai) [inline plugins/wasi_nn/piper.cpp:36]: Let's apply `string_view_literal` for these error messages. Use `"xxx"sv` instead.

**Review by hydai — CHANGES_REQUESTED** [review]
cc @dm4 and @q82419 PTAL.

**Comment by dm4** [comment]
I have some thoughts:
1. It looks like the implementation of the piper backend directly puts the json format config into `Builders[0]` and passes it into `load()`. I wonder if using this together with the current `--nn-preload` will cause any issues?
2. Is there a detailed option list available for the currently supported options of the piper backend?

**Comment by PeterD1524** [comment]
> 1. It looks like the implementation of the piper backend directly puts the json format config into `Builders[0]` and passes it into `load()`. I wonder if using this together with the current `--nn-preload` will cause any issues?

I'm not sure how `--nn-preload` works. Will passing something like `--nn-preload=default:Piper:AUTO:options.json` be ok?
Piper requires these files:
1. a .onnx model (required)
2. a config file (required)
3. a espeak-ng data directory (optional for espeak models, but only 1 of [the 124 officially supported voice models](https://github.com/rhasspy/piper/blob/master/VOICES.md) does not require this)
4. a libtashkeel ort model (optional, required for Arabic language)

> 2\. Is there a detailed option list available for the currently supported options of the piper backend?

https://github.com/WasmEdge/WasmEdge/blob/af998264e41e3e1efc55faee45c3fae107ad4ee4/plugins/wasi_nn/piper.h#L24-L67

**Comment by dm4** [comment]
> > 1. It looks like the implementation of the piper backend directly puts the json format config into `Builders[0]` and passes it into `load()`. I wonder if using this together with the current `--nn-preload` will cause any issues?
> 
> I'm not sure how `--nn-preload` works. Will passing something like `--nn-preload=default:Piper:AUTO:options.json` be ok?

Oh, I understand. That's no problem then.

> > 2. Is there a detailed option list available for the currently supported options of the piper backend?
> 
> https://github.com/WasmEdge/WasmEdge/blob/af998264e41e3e1efc55faee45c3fae107ad4ee4/plugins/wasi_nn/piper.h#L24-L67

Okay.

**Review by dm4 — APPROVED** [review]

**Comment by hydai** [comment]
Hi @PeterD1524 
Please fix the conflicts. I think this PR is ready to merge if all CI workflows pass.

File: plugins/wasi_nn/piper.cpp [hunk 3]
```diff
@@ -18,64 +20,71 @@
...
+  auto [Err, HasValue] = getOption(Object, Key, Value);
+  if (HasValue) {
+    Result = Value;
   }
-  return WASINN::ErrNo::Success;
+  return Err;
+}
+
+bool fileExists(const std::string &Filename) {
+  auto Stream = std::ifstream{Filename, std::ios::binary};
```
Comment (hydai) [inline plugins/wasi_nn/piper.cpp:58]: I have a question. What if the `Filename` is a utf8string? Can it be handled in this way?

**Review by hydai — CHANGES_REQUESTED** [review]

File: plugins/wasi_nn/piper.cpp (hunk 3 above)
Comment (PeterD1524) [reply plugins/wasi_nn/piper.cpp:56]: Do you mean if `Filename` is UTF-8 encoded? I guess it depends on whether the native filesystem encoding is also UTF-8 :(

Comment (hydai) [reply plugins/wasi_nn/piper.cpp:56]: How about using `std::filesystem::u8path` before passing to `std::ifstream`, which will ensure the u8string is handled correctly by the filesystem?

Comment (hydai) [reply plugins/wasi_nn/piper.cpp:56]: Just like we did in the filesystem related tests. Ref: https://github.com/WasmEdge/WasmEdge/blob/9e9974411f444f588ca0cdd3550349ccabda4cf2/test/api/APIAOTCoreTest.cpp#L59-L61

Comment (PeterD1524) [reply plugins/wasi_nn/piper.cpp]: done in https://github.com/WasmEdge/WasmEdge/pull/3499/commits/0d13802d0a1e177103d95bfa33340e3c332206b6

**Review by hydai — APPROVED** [review]
LGTM.

**Comment by hydai** [comment]
CleanShot 2024-07-11 at 23 14 33
GitHub doesn't allow me to merge. @PeterD1524 Could you please try to rebase to the latest master branch again?

**Comment by hydai** [comment]
Never mind. I fixed it. Thanks @PeterD1524

**Pull request merged**